File: Data Structures & Algorithms/copy-linked-list-with-random-pointer/submission-3.js
// class Node {
//   constructor(val, next = null, random = null) {
//       this.val = val;
//       this.next = next;
//       this.random = random;
//   }
// }

class Solution {
    /**
     * @param {Node} head
     * @return {Node}
     */
    copyRandomList(head) {
        if(!head) return head
        let map = new Map()
        let curr = head

        while(curr){
            map.set(curr, new Node(curr.val))
            curr = curr.next
        }
        curr = head
        while(curr){
            const cloned = map.get(curr)
            cloned.next = map.get(curr.next) ?? null
            cloned.random = map.get(curr.random) ?? null
            curr = curr.next
        }
        return map.get(head)
    }
}

File: Data Structures & Algorithms/reorder-linked-list/submission-3.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    void reorderList(ListNode* head) {
        ListNode* slow = head;
        ListNode* fast = head->next;

        while(fast && fast->next){
            fast = fast ->next->next;
            slow = slow->next;
        }
        ListNode* slowNext = slow->next;
        slow->next = NULL;
        ListNode* prev = NULL;
        ListNode* curr = slowNext;

        while(curr){
            ListNode* next = curr->next;
            curr->next = prev;
            prev = curr;
            curr = next;
        }
        curr = head;
        while(prev){
            ListNode *currNext = curr->next, *prevNext = prev->next;
            curr->next = prev;
            prev->next = currNext;
            curr = currNext;
            prev = prevNext;
        }
    }
};
